keps/sig-node/4328-node-label-manager/README.md

```diff
@@ -0,0 +1,404 @@
+`RequiredDuringSchedulingRequiredDuringExecution`  implies that pods with this affinity rule must meet certain conditions during both the scheduling phase and the execution phase.
+
+## Motivation
+- Enable strict node affinity for pod scheduling and execution.
+- Provide users with greater control over pod placement.
+- Ensure predictable and consistent pod behavior in various scenarios.
+- Address use cases with specific placement and execution requirements.
+  
+### Goals
+- Introduce the "RequiredDuringSchedulingRequiredDuringExecution" node affinity type.
+- Add a node affinity controller to the controllers to ensure pods being evicted if the selectors are no longer met at runtime.
+
+### Non-Goals
+- This KEP does not affect existing node affinity types.
+- It does not introduce breaking changes for workloads with different requirements.
+- It does not enforce stricter affinity rules for existing workloads.
+- It does not modify core Kubernetes components unrelated to node affinity.
+
+## Proposal
+- Implement the "RequiredDuringSchedulingRequiredDuringExecution" node affinity type.
+- Update the kube-scheduler, kube-controller-manager and kube-apiserver.
+- Define the behavior, syntax, and use cases in documentation.
+- Facilitate testing, adoption, and maintenance of this feature.
+
+### User Stories (Optional)
+
+#### Optimizing Workloads Migrating
+As a Data Scientist, I want to ensure that my GPU-intensive ml workloads are always on nodes with correct type of GPUs during both scheduling and execution. 
+So I labeled one node with "gpu=g1". I added some more powerful nodes, and now I want to move all pods requiring gpu to this new node. 
+With this feature, I can simply remove the "gpu=g1" from the first node and adds it to the new node, so I can move the target pods and keep the remaining pods intact.
+So far descheduler has the ability to achieve this goal. However, bringing a new component to the system introduces both a lot of uncertainty and additional learning costs. So as a user, I hope that kubernetes can natively support this feature so that users can do it without spending extra learning costs.
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
+
+The major concern for this controller may be the overhead to the nodes with a large number of pods. 
+Every time a node's label is updated, the controller needs to iterate all pods assigned to the node. This may cause potential scaling issues.
+
+## Design Details
+
+Add a node label manager similar to the existing taint_eviction controller, but under the node lifecycle controller.
+
+The manager do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the nodes, check the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, check if the selector still met to the new node label. If not, evict the pod.
```

Comment:
would this use the eviction API or direct Delete? would it respect PDB?

Reply:
I tend to delete directly, don't respect PDB.
`RequiredDuringExecution` is a strict affinity. I think users configure their pods with this affinity to indicate that this pod will not run properly on mismatched nodes. And in this case, there's no need to leave a pod as it will fail anyway.

But we should definitely respect `terminationGracePeriodSeconds`, as most use case of this is for migration.

However I looked into descheduler's code today, and descheduler seems always use the eviction API to respect PDB. So I don’t have a strong opinion on this either and I'm open for a discussion.

Reply:
I think we should match the behavior of the taint-based-eviction. If your workload shouldn't be disrupted, then you probably shouldn't be using RequiredDuringExecution.